File: day-12/python/main.py
import pathlib
from collections import deque
from dataclasses import dataclass


@dataclass
class Waypoint:
    relative_east_position: int = 10
    relative_north_position: int = 1

    def move_east(self, units: int):
        self.relative_east_position += units

    def move_west(self, units: int):
        self.relative_east_position -= units

    def move_north(self, units: int):
        self.relative_north_position += units

    def move_south(self, units: int):
        self.relative_north_position -= units

    def rotate_right_around_ship(self, degrees: int):
        for _ in range(degrees // 90):
            self.relative_east_position, self.relative_north_position = (
                self.relative_north_position,
                -self.relative_east_position,
            )

    def rotate_left_around_ship(self, degrees: int):
        for _ in range(degrees // 90):
            self.relative_east_position, self.relative_north_position = (
                -self.relative_north_position,
                self.relative_east_position,
            )


class Ship:
    def __init__(self, action_map_part: int):
        self.waypoint = Waypoint()
        self.cardinal_directions = deque(('E', 'N', 'W', 'S'))
        self.east_position = 0
        self.north_position = 0
        self.action_map_part = action_map_part
        self.action_maps = {
            1: {
                'N': self.move_north,
                'E': self.move_east,
                'S': self.move_south,
                'W': self.move_west,
                'L': self.turn_left,
                'R': self.turn_right,
                'F': self.move_forward,
            },
            2: {
                'N': self.waypoint.move_north,
                'E': self.waypoint.move_east,
                'S': self.waypoint.move_south,
                'W': self.waypoint.move_west,
                'L': self.waypoint.rotate_left_around_ship,
                'R': self.waypoint.rotate_right_around_ship,
                'F': self.move_toward_waypoint,
            },
        }

    @property
    def current_direction(self):
        return self.cardinal_directions[0]

    def move_east(self, units: int):
        self.east_position += units

    def move_west(self, units: int):
        self.east_position -= units

    def move_north(self, units: int):
        self.north_position += units

    def move_south(self, units: int):
        self.north_position -= units

    def turn_left(self, degrees: int):
        self.cardinal_directions.rotate(-degrees // 90)

    def turn_right(self, degrees: int):
        self.cardinal_directions.rotate(degrees // 90)

    def move_forward(self, units: int):
        self.action_maps[self.action_map_part][self.current_direction](units)

    def move_toward_waypoint(self, units: int):
        self.east_position += units * self.waypoint.relative_east_position
        self.north_position += units * self.waypoint.relative_north_position

    def process_instruction(self, action: str, value: int):
        self.action_maps[self.action_map_part][action](value)

    @property
    def manhattan_distance(self) -> int:
        return abs(self.east_position) + abs(self.north_position)


def compute_manhattan_distance(instructions: list[tuple[str, int]], part: int) -> int:
    ship = Ship(part)
    for action, value in instructions:
        ship.process_instruction(action, value)
    return ship.manhattan_distance


with open(pathlib.Path(__file__).parent.parent / 'input.txt') as f:
    instructions = [(line[0], int(line[1:])) for line in f.read().splitlines()]

part_one_solution = compute_manhattan_distance(instructions, 1)
part_two_solution = compute_manhattan_distance(instructions, 2)

print('Part One:', part_one_solution)
print('Part Two:', part_two_solution)

assert part_one_solution == 1106
assert part_two_solution == 107281

File: day-12/rust/src/main.rs
use std::collections::VecDeque;
use std::fs;
use std::mem;
use std::path::Path;

struct Waypoint {
    pub relative_east_position: i32,
    pub relative_north_position: i32,
}

impl Default for Waypoint {
    fn default() -> Self {
        Waypoint {
            relative_east_position: 10,
            relative_north_position: 1,
        }
    }
}

struct Ship {
    cardinal_directions: VecDeque<char>,
    east_position: i32,
    north_position: i32,
    waypoint: Waypoint,
    pub action_map_part: u32,
}

impl Default for Ship {
    fn default() -> Self {
        Ship {
            cardinal_directions: vec!['E', 'N', 'W', 'S'].into_iter().collect(),
            east_position: 0,
            north_position: 0,
            waypoint: Waypoint::default(),
            action_map_part: 0,
        }
    }
}

impl Ship {
    fn current_direction(&self) -> char {
        self.cardinal_directions[0]
    }

    fn manhattan_distance(&self) -> u32 {
        (self.east_position.abs() + self.north_position.abs()) as u32
    }

    fn process_instruction(&mut self, action: char, value: i32) {
        match &self.action_map_part {
            1 => match action {
                'N' => self.north_position += value,
                'S' => self.north_position -= value,
                'E' => self.east_position += value,
                'W' => self.east_position -= value,
                'R' => self.cardinal_directions.rotate_right((value / 90) as usize),
                'L' => self.cardinal_directions.rotate_left((value / 90) as usize),
                'F' => self.process_instruction(self.current_direction(), value),
                _ => panic!("unrecognized instruction '{}{}'", action, value),
            },
            2 => match action {
                'N' => self.waypoint.relative_north_position += value,
                'S' => self.waypoint.relative_north_position -= value,
                'E' => self.waypoint.relative_east_position += value,
                'W' => self.waypoint.relative_east_position -= value,
                'R' => {
                    for _ in 0..(value / 90) as usize {
                        mem::swap(
                            &mut self.waypoint.relative_east_position,
                            &mut self.waypoint.relative_north_position,
                        );
                        self.waypoint.relative_north_position =
                            -self.waypoint.relative_north_position;
                    }
                }
                'L' => {
                    for _ in 0..(value / 90) as usize {
                        mem::swap(
                            &mut self.waypoint.relative_east_position,
                            &mut self.waypoint.relative_north_position,
                        );
                        self.waypoint.relative_east_position =
                            -self.waypoint.relative_east_position;
                    }
                }
                'F' => {
                    self.east_position += value * self.waypoint.relative_east_position;
                    self.north_position += value * self.waypoint.relative_north_position;
                }
                _ => panic!("unrecognized instruction '{}{}'", action, value),
            },
            _ => unreachable!(),
        }
    }
}

fn compute_manhattan_distance(instructions: &[(char, i32)], part: u32) -> u32 {
    let mut ship = Ship {
        action_map_part: part,
        ..Ship::default()
    };
    for (action, value) in instructions.iter() {
        ship.process_instruction(*action, *value);
    }
    ship.manhattan_distance()
}

fn main() {
    let input = fs::read_to_string(
        Path::new(env!("CARGO_MANIFEST_DIR"))
            .parent()
            .unwrap()
            .join("input.txt"),
    )
    .unwrap();
    let instructions: Vec<_> = input
        .lines()
        .map(|l| {
            let (action, value) = l.split_at(1);
            (
                action.chars().next().unwrap(),
                value.parse::<i32>().unwrap(),
            )
        })
        .collect();

    let part_one_solution = compute_manhattan_distance(&instructions, 1);
    let part_two_solution = compute_manhattan_distance(&instructions, 2);

    println!("Part One: {}", part_one_solution);
    println!("Part Two: {}", part_two_solution);

    assert_eq!(part_one_solution, 1106);
    assert_eq!(part_two_solution, 107281);
}
